File: 二叉树/105.cpp
#include <vector>
#include <queue>
#include <string>
#include <unordered_map>
using namespace std;
//Definition for a binary tree node.
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    unordered_map<int, int> idx;
    int preIdx = 0;
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder)
    {

        int n = inorder.size();
        for (int i = 0; i < n; i++)
            idx[inorder[i]] = i;
        return build(preorder, inorder,0, n - 1);
    }

    TreeNode* build(vector<int>& preorder, vector<int>& inorder, int begin, int end)
    {
        if (begin<=end)
        {
            int val = preorder[preIdx];
            TreeNode* root = new TreeNode(val);
            int inIdx = idx[val];
            preIdx++;
            root->left = build(preorder, inorder, begin, inIdx-1);
            root->right = build(preorder, inorder,inIdx + 1, end);
            return root;
        }
        return nullptr;
    }
};
string treeNodeToString(TreeNode *root)
{
    if (root == nullptr)
    {
        return "[]";
    }

    string output = "";
    queue<TreeNode *> q;
    q.push(root);
    while (!q.empty())
    {
        TreeNode *node = q.front();
        q.pop();

        if (node == nullptr)
        {
            output += "null, ";
            continue;
        }

        output += to_string(node->val) + ", ";
        q.push(node->left);
        q.push(node->right);
    }
    return "[" + output.substr(0, output.length() - 2) + "]";
}

File: 二叉树/105.py
from typing import List


def treeNodeToString(root):
    if not root:
        return "[]"
    output = ""
    queue = [root]
    current = 0
    while current != len(queue):
        node = queue[current]
        current = current + 1

        if not node:
            output += "null, "
            continue

        output += str(node.val) + ", "
        queue.append(node.left)
        queue.append(node.right)
    return "[" + output[:-2] + "]"
# Definition for a binary tree node.


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        if(inorder):
            val = preorder.pop(0)
            root = TreeNode(val)
            idx = inorder.index(val)
            root.left = self.buildTree(preorder, inorder[:idx])
            root.right = self.buildTree(preorder, inorder[idx+1:])
            return root
        else:
            return None

preorder = [3,9,20,15,7]
inorder = [9,3,15,20,7]

obj = Solution()
root = obj.buildTree(preorder,inorder)

print(treeNodeToString(root))